### PR TITLE
nixos/yazi: support plugins and flavors

### PR DESCRIPTION
**Pull request by linsui**
## Description of changes

And fix the crash.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

@sxyazi

### CODE REVIEW

**Comment by sxyazi** [comment]
Thanks for the fix!

@XYenon can you take a look at this PR?

**Comment by teto** [comment]
I would prefer to see plugin support in pkgs/by-name/ya/yazi/package.nix, modules should be kind of last resort IMO. The module could keep the extensions but the implementation should live in the wrapper. 
It would makes the features accessible outside of nixos (includign home-manager)

**Comment by linsui** [comment]
Do you want to replace the module with a wrapper? Why is a wrapper preferred? I thought the wrapper will just become a mess. Those args in the wrapper can't be documented and discovered and no one knows how to use them without reading the code (I have seen enough examples). HM should implement their own module with plugins from nixpkgs.

**Comment by teto** [comment]
> Do you want to replace the module with a wrapper? 

Not necessarily, we can have both if you maintain the module.

> Why is a wrapper preferred? 

Because it can be used outside nixos.

> Those args in the wrapper can't be documented and discovered and no one knows how to use them without reading the code

says who ? of course they can be documented, just look at the nixpkgs documentation. Not only that but it's a lot easier to test a wrapper than a module.

**Comment by linsui** [comment]
> says who ? of course they can be documented, just look at the nixpkgs documentation. Not only that but it's a lot easier to test a wrapper than a module.

Right, the only problem is that only few of them are documented because it's not generated automatically.

> Because it can be used outside nixos.

Do you have any idea how to implement it? I thought we have to set YAZI_CONFIG_HOME in the wrapper and this overrides the default config files. So if the plugins are installed with the wrapper then the config also needs to be set with the wrapper. And this module will become a wrapper of the wrapper? Or a generator of the config home directory?

**Comment by teto** [comment]
> I thought we have to set YAZI_CONFIG_HOME in the wrapper and this overrides the default config files. So if the plugins are installed with the wrapper then the config also needs to be set with the wrapper.
 
This could be seen as an advantage, i.e, you get a pure config. but I recognize this can be a bother. Isn't it the same with this PR ? aka if you set files in YAZI_CONFIG_HOME = /etc/yazi then yazi will ignore the files in $XDG_CONFIG_HOME/yazi ?

There is a similar issue with the neovim wrapper.
The wrapper could (in impure mode, if we ever want that) generate an init.lua file that chain loads the one in $XDG_CONFIG_HOME ?

**Comment by linsui** [comment]
> aka if you set files in YAZI_CONFIG_HOME = /etc/yazi then yazi will ignore the files in $XDG_CONFIG_HOME/yazi ?

Yazi by default only loads the config files in home.

> The wrapper could (in impure mode, if we ever want that) generate an init.lua file that chain loads the one in $XDG_CONFIG_HOME ?

I don't think that yazi can load a config file in a config file...

OK, I know how to do that now.

**Comment by linsui** [comment]
I find a stupid problem. In fact I don't use the wrapper at all because I don't use those tools in the annoying wrapper. If I use the wrapper I have to set all of them to false. If the config is added to the wrapper I have to use the wrapper. This provide no benefits to me. So I add another option to disable all of them.

**Comment by linsui** [comment]
@teto Is there anything I can do? :)

File: pkgs/by-name/ya/yazi/package.nix [hunk 1]
```diff
@@ -34,7 +41,40 @@ let
...
+        ${lib.concatMapStringsSep
+        "\n"
+        (lib.mapAttrsToList (name: value: "ln -s ${value} $out/plugins/${name}") plugins)}
+    ''}
+
+    mkdir $out/flavors
+    ${lib.optionalString (flavors != { }) ''
+        ${lib.concatMapStringsSep
+        "\n"
+        (lib.mapAttrsToList (name: value: "ln -s ${value} $out/plugins/${name}") flavors)}
```
Comment (XYenon) [inline pkgs/by-name/ya/yazi/package.nix:70]:
```suggestion
        (lib.mapAttrsToList (name: value: "ln -s ${value} $out/flavors/${name}") flavors)}
```

Comment (linsui) [reply pkgs/by-name/ya/yazi/package.nix]: Good catch, thanks!

**Review by XYenon — APPROVED** [review]

**Pull request merged**